lib/galaxy/jobs/__init__.py

```diff
@@ -801,7 +803,10 @@ def job_destination(self):
 
         :returns: ``JobDestination``
         """
-        return self.job_runner_mapper.get_job_destination(self.params)
+        dest = self.job_runner_mapper.get_job_destination(self.params)
+        if self.dataset_path_rewriter is None:
+            self.dataset_path_rewriter = self._job_dataset_path_rewriter(self.working_directory)
+        return dest
```

Comment:
Are we sure that `self.dataset_path_rewriter` is never used before `self.job_destination` (e.g. in the `TaskWrapper` subclass)?
Alternatively I think we can move its initialisation right after `self.job_runner_mapper.cached_job_destination = JobDestination(from_job=job)` around line 728 above.

Reply:
Initially I had it in the `__init__` method, but it doesn't do any good to set it up there if the destination has not been resolved/cached yet. But yes, it does look like the `TaskWrapper` would use it before a destination is set. I don't think it would cause a failure in this case, it would simply fail to retrieve config params from the destination (as it did prior to this PR) for tasks, which is not ideal but not worse than the current situation. Do you think it's important to address this for tasks in this PR?

Reply:
> Initially I had it in the __init__ method, but it doesn't do any good to set it up there if the destination has not been resolved/cached yet.

I see that can be an issue when `use_persisted_destination` is `False`, what about converting `dataset_path_rewriter` from an attribute to a property (which should be easy when removing its only parameter `self.working_directory`)?